### ISSUE TITLE
Systemd-networkd bridge creation fails on ARM (ODROID-C2) but works on x86_64

### ISSUE THREAD

**Issue by graysky2** [systemd/systemd#4945]
### Submission type

  - [ X ] Bug report

### systemd version the issue has been seen with

> 232-6

### Used distribution

> Arch ARM on ODROID C2

### In case of bug report: Expected behaviour you didn't see

> Bridge networking should come up.

### In case of bug report: Unexpected behaviour you saw

> Bridge networking dies and cannot start.

### In case of bug report: Steps to reproduce the problem

1. I have bridge-utils installed.
2. I have systemd-networkd and systemd-resolved enabled.
3. I setup the following 3 files in /etc/systemd/network

```
/etc/systemd/network/br0.network 
[Match]
Name=br0
[Network]
DHCP=yes
```

```
/etc/systemd/network/br0.netdev
[NetDev]
Name=br0
Kind=bridge
```

```
/etc/systemd/network/uplink.netdev
[Match]
Name=eth0
[Network]
Bridge=br0
```

When the C2 boots, however, I get the following in the journalctl and no functional network:

```
% journalctl -b -u systemd-networkd
-- Logs begin at Fri 2016-12-09 20:10:44 EST, end at Wed 2016-12-21 12:32:01 EST. --
Dec 21 12:19:55 odroid systemd[1]: Starting Network Service...
Dec 21 12:19:55 odroid systemd-networkd[227]: br0: netdev ready
Dec 21 12:19:55 odroid systemd-networkd[227]: Enumeration completed
Dec 21 12:19:55 odroid systemd[1]: Started Network Service.
Dec 21 12:19:55 odroid systemd-networkd[227]: br0: IPv6 enabled for interface: Success
Dec 21 12:19:55 odroid systemd-networkd[227]: br0: Bridge parameters could not be set: Operation not supported
Dec 21 12:19:55 odroid systemd-networkd[227]: br0: Could not bring up interface: Invalid argument
Dec 21 12:19:55 odroid systemd-networkd[227]: eth0: Could not join netdev: Invalid argument
Dec 21 12:19:55 odroid systemd-networkd[227]: eth0: Failed
```

If I use a simple DHCP setup, the network works fine with systemd-netctld:
```
/etc/systemd/network/eth0.network 
[Match]
Name=eth0
[Network]
DHCP=yes
```

**Comment by poettering** [comment]
are you sure you you have bridge support enabled properly and comprehensively in the kernel on both systems? This really looks like your kernel on the system doesn't support bridging networkd wants to set it up.

Which kernel versions are these?

**Comment by graysky2** [comment]
> Which kernel versions are these?

The ODROID-C2 is running the latest supported kernel: 3.14.79 and my x86_64 box is using 3.9.0 currently but the functional behavior has been there for some time (perhaps back to the 4.4 series).

> are you sure you you have bridge support enabled properly and comprehensively in the kernel on both systems?

I believe it is correctly configured.  I just now discovered that if I boot using the simple DHCP netowkr ("eth0.network" shown above) and THEN delete the "eth0.network" file and replace it with the 3 needed for the bridge, a restart of systemd-networkd brings up the bridge jsut fine.  BUT, this only happens if I first boot with the simple DHCP setup, then preform file moves.

Here is a detailed list of what I did and some journactl output:

1. Have just /etc/systemd/network/eth0.network present.
2. Reboot.  Network comes up as expected.

```
% sudo journalctl -b -u systemd-networkd
-- Logs begin at Fri 2016-12-09 20:10:44 EST, end at Wed 2016-12-21 12:50:12 EST. --
Dec 21 12:50:00 odroid systemd[1]: Starting Network Service...
Dec 21 12:50:00 odroid systemd-networkd[217]: Enumeration completed
Dec 21 12:50:00 odroid systemd-networkd[217]: eth0: IPv6 enabled for interface: Success
Dec 21 12:50:00 odroid systemd[1]: Started Network Service.
Dec 21 12:50:00 odroid systemd-networkd[217]: eth0: Gained carrier
Dec 21 12:50:01 odroid systemd-networkd[217]: eth0: Could not bring up interface: Invalid argument
Dec 21 12:50:02 odroid systemd-networkd[217]: eth0: Gained IPv6LL
Dec 21 12:50:02 odroid systemd-networkd[217]: eth0: Lost carrier
Dec 21 12:50:04 odroid systemd-networkd[217]: eth0: Gained carrier
Dec 21 12:50:04 odroid systemd-networkd[217]: eth0: DHCPv4 address 192.168.1.242/24 via 192.168.1.1
Dec 21 12:50:13 odroid systemd-networkd[217]: eth0: Configured
```

3. Remove eth0.network from `/etc/systemd/network/` and replace it with the 3 files above (br0.network, br0.netdev, and uplink.network).
4. Restart systemd-networkd and everything works as expected.

```
% sudo journalctl -b -u systemd-networkd
Dec 21 12:52:42 odroid systemd[1]: Stopping Network Service...
Dec 21 12:52:42 odroid systemd[1]: Stopped Network Service.
Dec 21 12:52:42 odroid systemd[1]: Starting Network Service...
Dec 21 12:52:42 odroid systemd-networkd[460]: br0: netdev ready
Dec 21 12:52:42 odroid systemd-networkd[460]: eth0: Gained IPv6LL
Dec 21 12:52:42 odroid systemd-networkd[460]: Enumeration completed
Dec 21 12:52:42 odroid systemd[1]: Started Network Service.
Dec 21 12:52:42 odroid systemd-networkd[460]: br0: Bridge parameters could not be set: Operation not supported
Dec 21 12:52:42 odroid systemd-networkd[460]: br0: IPv6 enabled for interface: Success
Dec 21 12:52:42 odroid systemd-networkd[460]: br0: Could not bring up interface: Invalid argument
Dec 21 12:52:42 odroid systemd-networkd[460]: br0: Gained carrier
Dec 21 12:52:42 odroid systemd-networkd[460]: br0: DHCPv4 address 192.168.1.209/24 via 192.168.1.1
Dec 21 12:52:44 odroid systemd-networkd[460]: br0: Gained IPv6LL
Dec 21 12:52:55 odroid systemd-networkd[460]: eth0: Configured
Dec 21 12:52:57 odroid systemd-networkd[460]: br0: Configured
```

I cannot explain this.  Again, if I boot directly with the 3 files for the bridge, the journalctl output is as in my first report and there is no functional network.

**Comment by poettering** [comment]
Those kernels are pretty old. I am pretty sure this is some kernel issue where module auto-loading doesn't work they way it does these days. When you got it running, do an "lsmod" and add all possibly bridge-related kernel modules to /etc/load-modules.d/… and make sure this is ordered before systemd-networkd.service...

**Comment by graysky2** [comment]
That is a good guess.  Here is what I did:
1. Boot with the simple network profile.
2. `lsmod > first`
3. Did what I described above to get the bridge working.
4. `lsmod > second`

The thing is that there is no difference between the two lsmod outputs....

```
% cat first
Module                  Size  Used by
ip6table_filter         1433  0
ip6_tables             16563  1 ip6table_filter
iptable_filter          1485  0
spidev                  6163  0
spi_gpio                6410  0
spi_bitbang             3695  1 spi_gpio
w1_gpio                 3553  0
wire                   24167  1 w1_gpio
gxbb_wdt                3043  0
sch_fq_codel            7257  2
ip_tables              16493  1 iptable_filter
autofs4                26431  2
```

```
% cat second
Module                  Size  Used by
ip6table_filter         1433  0
ip6_tables             16563  1 ip6table_filter
iptable_filter          1485  0
spidev                  6163  0
spi_gpio                6410  0
spi_bitbang             3695  1 spi_gpio
w1_gpio                 3553  0
wire                   24167  1 w1_gpio
gxbb_wdt                3043  0
sch_fq_codel            7257  2
ip_tables              16493  1 iptable_filter
autofs4                26431  2
```

**Comment by graysky2** [comment]
@poettering - I tried booting 4.9.0 on the ODROID-C2 but the hardware support is just not there yet (would not boot).  The next best experiment I thought of would be to build and boot into 3.14.79 on my x86_64 box.  I am typing this now from 3.14.79 on x86_64 with the same systemd-networkd setup as on the ODROID-C2 and have experienced no such problems with the bridge.  Not sure what is to blame on the ARM.

```
% uname -r
3.14.79-1-lts

% journalctl -b -u systemd-networkd
-- Logs begin at Sat 2016-12-17 09:40:18 EST, end at Wed 2016-12-21 17:00:02 EST. --
Dec 21 16:58:10 ease systemd[1]: Starting Network Service...
Dec 21 16:58:10 ease systemd-networkd[517]: br0: netdev ready
Dec 21 16:58:10 ease systemd-networkd[517]: Enumeration completed
Dec 21 16:58:10 ease systemd[1]: Started Network Service.
Dec 21 16:58:10 ease systemd-networkd[517]: br0: IPv6 enabled for interface: Success
Dec 21 16:58:10 ease systemd-networkd[517]: br0: Bridge parameters could not be set: Operation not supported
Dec 21 16:58:10 ease systemd-networkd[517]: br0: Could not bring up interface: Invalid argument
Dec 21 16:58:10 ease systemd-networkd[517]: eth0: IPv6 disabled for interface: Success
Dec 21 16:58:14 ease systemd-networkd[517]: eth0: Gained carrier
Dec 21 16:58:14 ease systemd-networkd[517]: br0: Gained carrier
Dec 21 16:58:14 ease systemd-networkd[517]: br0: DHCPv4 address 192.168.1.201/24 via 192.168.1.1
Dec 21 16:58:15 ease systemd-networkd[517]: br0: Gained IPv6LL
Dec 21 16:58:28 ease systemd-networkd[517]: br0: Configured
```

**Comment by poettering** [comment]
Either way, it's the kernel which responds with an error here. I seriously doubt this has anything to do with systemd, this looks very much like a kernel problem of some kind, that likely is fixed with a newer kernel.

**Comment by graysky2** [comment]
If it is a kernel issue, I would think that it would also happen on the x86_64 box using the same kernel version, no?  I can successfully use netctl to create the bridge on the ODROID.  Seems like something related to systemd-networkd...

**Comment by poettering** [comment]
well, who knows, might be race, or something. A much more interesting test would be to use a current kernel on your odroid, and see if that fixes it...

**Comment by graysky2** [comment]
I agree.  That is the killer experiment to validate your hypothesis.  Problem is that the current does not boot.  I have tried building from the linux-next tree but encountered errors as well.

**Comment by poettering** [comment]
So, I hope it's OK if I close this bug here. I am very sure this is just an issue triggered by networkd, but not in networkd, and there's a good chance this already got fixed in the current kernel. I understand that a newer kernel is not an option for you at the moment, but still, it's a kernel issue I am sure.

Hence, let's close this here for now. Please reopen if this fails on current kernels too...

Sorry, but I hope this makes sense.

**Comment by graysky2** [comment]
@poettering - I have this solved (workaround).  What I did was to make a oneshot service that runs a tiny script that setups up the bridge.  I then point systemd-networkd to the bridge and it works.  I searched through the systemd.network man page looking to see if there is a way I can have that make the bridge (run some code like a pre-up) but didn't see anything... is there?

EDIT: https://github.com/graysky2/odroid-auto-bridge

Thanks again for all the work you do with systemd.
